File: java/src/main/java/org/revenj/postgres/PostgresDatabaseNotification.java
package org.revenj.postgres;

import org.postgresql.PGNotification;
import org.postgresql.core.BaseConnection;
import org.revenj.Utils;
import org.revenj.patterns.*;
import org.revenj.postgres.converters.StringConverter;
import rx.Observable;
import rx.subjects.PublishSubject;

import java.io.Closeable;
import java.io.IOException;
import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.function.Function;

public class PostgresDatabaseNotification implements EagerNotification, Closeable {

	private final Function<ServiceLocator, Connection> connectionFactory;
	private final Optional<DomainModel> domainModel;
	private final ServiceLocator locator;

	private final PublishSubject<NotifyInfo> subject = PublishSubject.create();
	private final Observable<NotifyInfo> notifications;

	private final ConcurrentMap<Class<?>, Repository> repositories = new ConcurrentHashMap<>();
	private final ConcurrentMap<String, HashSet<Class<?>>> targets = new ConcurrentHashMap<>();

	private int retryCount;

	private Connection connection;
	private boolean isClosed;

	public PostgresDatabaseNotification(Function<ServiceLocator, Connection> connectionFactory, Optional<DomainModel> domainModel, ServiceLocator locator) {
		this.connectionFactory = connectionFactory;
		this.domainModel = domainModel;
		this.locator = locator;
		notifications = subject.asObservable();
		setupConnection();
		Runtime.getRuntime().addShutdownHook(new Thread(() -> isClosed = true));
	}

	private void setupConnection() {
		retryCount++;
		if (retryCount > 60) {
			retryCount = 30;
		}
		try {
			if (connection != null) {
				try {
					connection.close();
				} catch (Exception ex) {
					ex.printStackTrace();
				}
			}
			connection = connectionFactory.apply(locator);
			if (connection instanceof BaseConnection == false) return;
			BaseConnection bc = (BaseConnection) connection;
			Statement stmt = bc.createStatement();
			stmt.execute("LISTEN events; LISTEN aggregate_roots");
			retryCount = 0;
			Pooling pooling = new Pooling(bc, stmt);
			new Thread(pooling).start();
		} catch (Exception ex) {
			try {
				Thread.sleep(1000 * retryCount);
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
	}

	class Pooling implements Runnable {
		private final BaseConnection connection;
		private final Statement ping;

		public Pooling(BaseConnection connection, Statement ping) {
			this.connection = connection;
			this.ping = ping;
		}

		@Override
		public void run() {
			PostgresReader reader = new PostgresReader();
			while (!isClosed) {
				try {
					ping.execute(";");
					PGNotification[] notifications = connection.getNotifications();
					if (notifications == null || notifications.length == 0) {
						try {
							Thread.sleep(500);
						} catch (InterruptedException e) {
							e.printStackTrace();
						}
						continue;
					}
					for (PGNotification n : notifications) {
						if (!"events".equals(n.getName()) && !"aggregate_roots".equals(n.getName())) continue;
						String param = n.getParameter();
						String ident = param.substring(0, param.indexOf(':'));
						String op = param.substring(ident.length() + 1, param.indexOf(':', ident.length() + 1));
						String values = param.substring(ident.length() + op.length() + 2);
						reader.process(values);
						List<String> ids = StringConverter.parseCollection(reader, 0, false);
						if (ids != null && ids.size() > 0) {
							String[] uris = ids.toArray(new String[ids.size()]);
							switch (op) {
								case "Update":
									subject.onNext(new NotifyInfo(ident, Operation.Update, uris));
									break;
								case "Change":
									subject.onNext(new NotifyInfo(ident, Operation.Change, uris));
									break;
								case "Delete":
									subject.onNext(new NotifyInfo(ident, Operation.Delete, uris));
									break;
								default:
									subject.onNext(new NotifyInfo(ident, Operation.Insert, uris));
									break;
							}
						}
					}
				} catch (SQLException | IOException ex) {
					try {
						Thread.sleep(1000);
					} catch (InterruptedException e) {
						e.printStackTrace();
					}
					setupConnection();
					return;
				}
			}
		}
	}

	private Repository getRepository(Class<?> manifest) {
		return repositories.computeIfAbsent(manifest, clazz ->
		{
			try {
				return (Repository) locator.resolve(Utils.makeGenericType(Repository.class, manifest));
			} catch (ReflectiveOperationException ex) {
				throw new RuntimeException("Repository is not registered for: " + manifest, ex);
			}
		});
	}

	@Override
	public void notify(NotifyInfo info) {
		subject.onNext(info);
	}

	@Override
	public Observable<NotifyInfo> getNotifications() {
		return notifications;
	}

	@Override
	public <T> Observable<TrackInfo<T>> track(final Class<T> manifest) {
		return notifications.filter(it -> {
			HashSet<Class<?>> set = targets.get(it.name);
			if (set == null) {
				set = new HashSet<>();
				Optional<Class<?>> domainType = domainModel.get().find(it.name);
				if (domainType.isPresent()) {
					set.add(domainType.get());
					Collections.addAll(set, domainType.get().getInterfaces());
				}
				targets.put(it.name, set);
			}
			return set.contains(manifest);
		}).map(it -> new TrackInfo<T>(it.uris, () -> getRepository(manifest).find(it.uris)));
	}

	public void close() {
		if (isClosed) {
			return;
		}
		isClosed = true;
		try {
			if (connection != null && !connection.isClosed()) {
				connection.close();
			}
		} catch (SQLException e) {
			e.printStackTrace();
		}
		connection = null;
	}
}

File: csharp/Core/Revenj.Core/DatabasePersistence/Postgres/PostgresDatabaseNotification.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Configuration;
using System.Data;
using System.Diagnostics;
using System.Diagnostics.Contracts;
using System.Reactive.Linq;
using System.Reactive.Subjects;
using System.Threading;
using Revenj.DatabasePersistence.Postgres.Converters;
using Revenj.DatabasePersistence.Postgres.Npgsql;
using Revenj.DomainPatterns;
using Revenj.Extensibility;
using Revenj.Utility;

namespace Revenj.DatabasePersistence.Postgres
{
	internal class PostgresDatabaseNotification : IEagerNotification, IDisposable
	{
		private static readonly TraceSource TraceSource = new TraceSource("Revenj.Database");

		private NpgsqlConnection Connection;
		private readonly Subject<NotifyInfo> Subject = new Subject<NotifyInfo>();
		private bool IsDisposed;
		private readonly Lazy<IDomainModel> DomainModel;
		private readonly ConcurrentDictionary<string, HashSet<Type>> Targets = new ConcurrentDictionary<string, HashSet<Type>>(1, 17);
		private int RetryCount;
		private readonly ConcurrentDictionary<Type, IRepository<IIdentifiable>> Repositories =
			new ConcurrentDictionary<Type, IRepository<IIdentifiable>>(1, 17);
		private readonly IServiceProvider Locator;
		private readonly BufferedTextReader Reader = new BufferedTextReader(string.Empty, new char[64], new char[8192]);
		private readonly ISystemState SystemState;

		public PostgresDatabaseNotification(
			ConnectionInfo connectionInfo,
			Lazy<IDomainModel> domainModel,
			ISystemState systemState,
			IServiceProvider locator)
		{
			Contract.Requires(connectionInfo != null);
			Contract.Requires(domainModel != null);
			Contract.Requires(locator != null);
			Contract.Requires(systemState != null);

			this.DomainModel = domainModel;
			this.Locator = locator;
			this.SystemState = systemState;
			Notifications = Subject.AsObservable();
			if (ConfigurationManager.AppSettings["Revenj.Notifications"] != "disabled")
				SetUpConnection(connectionInfo.ConnectionString + ";SyncNotification=true");
			AppDomain.CurrentDomain.ProcessExit += (s, ea) => IsDisposed = true;
			AppDomain.CurrentDomain.DomainUnload += (s, ea) => IsDisposed = true;
		}

		private void SetUpConnection(string connectionString)
		{
			RetryCount++;
			if (RetryCount > 60)
			{
				TraceSource.TraceEvent(TraceEventType.Critical, 5130, "Retry count exceeded: {0}", connectionString);
				RetryCount = 30;
			}
			try
			{
				if (Connection != null)
				{
					Connection.StateChange -= Connection_StateChange;
					Connection.Notification -= Connection_Notification;
					try { Connection.Close(); }
					catch (Exception ex)
					{
						TraceSource.TraceEvent(TraceEventType.Error, 5132, "{0}", ex);
					}
					try { Connection.Dispose(); }
					catch (Exception ex)
					{
						TraceSource.TraceEvent(TraceEventType.Error, 5133, "{0}", ex);
					}
				}
				Connection = new NpgsqlConnection(connectionString);
				Connection.StateChange += Connection_StateChange;
				Connection.Notification += Connection_Notification;
				Connection.Open();
				var com = Connection.CreateCommand();
				com.CommandText = "listen events; listen aggregate_roots; listen migration;";
				com.ExecuteNonQuery();
				RetryCount = 0;
			}
			catch (Exception ex)
			{
				TraceSource.TraceEvent(TraceEventType.Error, 5134, "{0}", ex);
				Thread.Sleep(1000 * RetryCount);
			}
		}

		private void Connection_StateChange(object sender, StateChangeEventArgs e)
		{
			if (IsDisposed)
				return;
			if (e.CurrentState == ConnectionState.Closed
				|| e.CurrentState == ConnectionState.Broken)
				SetUpConnection(Connection.ConnectionString);
			else if (e.CurrentState != ConnectionState.Open)
			{
				TraceSource.TraceEvent(TraceEventType.Error, 5136, "Invalid notification state: {0}", e.CurrentState);
			}
		}

		public void Notify(NotifyInfo info) { Subject.OnNext(info); }

		private void Connection_Notification(object sender, NpgsqlNotificationEventArgs e)
		{
			try
			{
				if (e.Condition == "events" || e.Condition == "aggregate_roots")
				{
					TraceSource.TraceEvent(TraceEventType.Verbose, 5137, "Postgres notification: {0} with {1}", e.Condition, e.AdditionalInformation);
					var firstSeparator = e.AdditionalInformation.IndexOf(':');
					var name = e.AdditionalInformation.Substring(0, firstSeparator);
					var secondSeparator = e.AdditionalInformation.Substring(firstSeparator + 1).IndexOf(':');
					var op = e.AdditionalInformation.Substring(firstSeparator + 1, secondSeparator).Trim();
					var array = e.AdditionalInformation.Substring(firstSeparator + secondSeparator + 2).Trim();
					if (array.Length > 0)
					{
						var uris = StringConverter.ParseCollection(Reader.Reuse(array), 0, false).ToArray();
						switch (op)
						{
							case "Update":
								Subject.OnNext(new NotifyInfo(name, NotifyInfo.OperationEnum.Update, NotifyInfo.SourceEnum.Database, uris));
								break;
							case "Change":
								Subject.OnNext(new NotifyInfo(name, NotifyInfo.OperationEnum.Change, NotifyInfo.SourceEnum.Database, uris));
								break;
							case "Delete":
								Subject.OnNext(new NotifyInfo(name, NotifyInfo.OperationEnum.Delete, NotifyInfo.SourceEnum.Database, uris));
								break;
							default:
								Subject.OnNext(new NotifyInfo(name, NotifyInfo.OperationEnum.Insert, NotifyInfo.SourceEnum.Database, uris));
								break;
						}
					}
				}
				else if (e.Condition == "migration")
				{
					TraceSource.TraceEvent(TraceEventType.Information, 5155, "Postgres migration detected: {0} ", e.AdditionalInformation);
					SystemState.Notify(new SystemEvent("migration", e.AdditionalInformation));
				}
			}
			catch (Exception ex)
			{
				TraceSource.TraceEvent(TraceEventType.Error, 5138, "{0}{1} {2}", e.Condition, e.AdditionalInformation, ex);
			}
		}

		public IObservable<NotifyInfo> Notifications { get; private set; }

		private IRepository<IIdentifiable> GetRepository<T>(string name)
		{
			IRepository<IIdentifiable> repository;
			if (!Repositories.TryGetValue(typeof(T), out repository))
			{
				var source = DomainModel.Value.Find(name);
				repository = Locator.Resolve<IRepository<IIdentifiable>>(typeof(IRepository<>).MakeGenericType(source));
				Repositories.TryAdd(typeof(T), repository);
			}
			return repository;
		}

		public IObservable<KeyValuePair<string[], Lazy<T[]>>> Track<T>()
		{
			//TODO: notifications can lag and when they do, scope can be disposed
			var type = typeof(T);
			return
				Notifications.Where(it =>
				{
					HashSet<Type> set;
					if (!Targets.TryGetValue(it.Name, out set))
					{
						set = new HashSet<Type>();
						var domainType = DomainModel.Value.Find(it.Name);
						if (domainType != null)
						{
							set.Add(domainType);
							foreach (var i in domainType.GetInterfaces())
								set.Add(i);
						}
						Targets.TryAdd(it.Name, set);
					}
					return set.Contains(type);
				}).Select(it => new KeyValuePair<string[], Lazy<T[]>>(it.URI, new Lazy<T[]>(() => GetRepository<T>(it.Name).Find(it.URI) as T[])));
		}

		public void Dispose()
		{
			if (IsDisposed)
				return;
			IsDisposed = true;
			try
			{
				if (Connection != null && Connection.State == ConnectionState.Open)
					Connection.Close();
			}
			catch (Exception ex)
			{
				TraceSource.TraceEvent(TraceEventType.Error, 5139, "{0}", ex);
			}
			try { Connection.Dispose(); }
			catch (Exception ex)
			{
				TraceSource.TraceEvent(TraceEventType.Error, 5140, "{0}", ex);
			}
			Connection = null;
		}
	}
}
